forui/lib/src/widgets/pagination/pagination.dart

```diff
@@ -0,0 +1,216 @@
+        ),
+      ),
+    );
+
+    return ListenableBuilder(
+      listenable: controller,
+      builder: (context, _) {
+        final (start, end) = controller.calculateSiblingRange();
+
+        return Row(
+          mainAxisAlignment: MainAxisAlignment.center,
+          children: [
+            previous,
+            if (controller.page > controller.minPagesDisplayedAtEdges) ...[
+              if (controller.showEdges)
+                FPaginationItemData(page: 0, style: style, controller: controller, child: const _Page()),
+              elipsis,
+            ],
+            for (int i = start; i <= end; i++)
+              FPaginationItemData(page: i, style: style, controller: controller, child: const _Page()),
+            if (controller.page < (lastPage - controller.minPagesDisplayedAtEdges)) ...[
+              elipsis,
+              if (controller.showEdges)
+                FPaginationItemData(page: lastPage, style: style, controller: controller, child: const _Page()),
+            ],
+            next,
+          ],
+        );
+      },
+    );
+  }
+}
+
+@internal
+class FPaginationItemData extends InheritedWidget {
+  @useResult
+  static FPaginationItemData of(BuildContext context) {
+    final data = context.dependOnInheritedWidgetOfExactType<FPaginationItemData>();
+    assert(data != null, 'No FPaginationItemData found in context.');
+    return data!;
+  }
+
+  final int page;
+  final FPaginationController controller;
+  final FPaginationStyle style;
+
+  const FPaginationItemData({
+    required this.page,
+    required this.controller,
+    required this.style,
+    required super.child,
+    super.key,
+  });
+
+  @override
+  bool updateShouldNotify(covariant FPaginationItemData old) =>
+      page != old.page || controller != old.controller || style != old.style;
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties
+      ..add(IntProperty('page', page))
+      ..add(DiagnosticsProperty('controller', controller))
+      ..add(DiagnosticsProperty('style', style));
+  }
+}
+
+@internal
+class Action extends StatelessWidget {
+  final VoidCallback onPress;
+  final FPaginationStyle style;
+  final Widget child;
+
+  const Action({required this.onPress, required this.style, required this.child, super.key});
+
+  Action.previous({required this.style, required this.onPress, super.key})
+    : child = FIcon(FAssets.icons.chevronLeft, color: style.iconStyle.color, size: style.iconStyle.size);
+
+  Action.next({required this.onPress, required this.style, super.key})
+    : child = FIcon(FAssets.icons.chevronRight, color: style.iconStyle.color, size: style.iconStyle.size);
+
+  @override
+  Widget build(BuildContext context) => Padding(
+    padding: style.itemPadding,
+    child: FTappable(
+      semanticLabel: 'Pagination action',
```

Comment:
This semantic label isn't ideal. It should be more descriptive and mention what the widget actually does, i.e. "Previous" or "Next". It should also ideally be localized but we can mark it with a TODO for now.